File: Astar.py
def heuristic(x, y, gx, gy):
    # Manhattan distance
    return abs(x - gx) + abs(y - gy)

def a_star(grid, sx, sy, gx, gy):
    x, y = sx, sy

    while (x, y) != (gx, gy):
        print(f"({x},{y}) -> ", end="")

        if x < gx and grid[x + 1][y] != 1:
            x += 1
        elif x > gx and grid[x - 1][y] != 1:
            x -= 1
        elif y < gy and grid[x][y + 1] != 1:
            y += 1
        elif y > gy and grid[x][y - 1] != 1:
            y -= 1
        else:
            print("Blocked! No path found.")
            return

    print(f"({gx},{gy})")

# 0 = empty cell, 1 = obstacle
grid = [
    [0, 0, 0],
    [0, 1, 0],  # obstacle in the middle
    [0, 0, 0]
]

a_star(grid, 0, 0, 2, 2)

File: Astar.cpp
#include <iostream>
#include <cmath>
using namespace std;

// Heuristic function (Manhattan distance)
int heuristic(int x, int y, int gx, int gy)
{
    return abs(x - gx) + abs(y - gy);
}

// Simplified A* simulation for a small grid
void aStar(int grid[3][3], int sx, int sy, int gx, int gy)
{
    int x = sx, y = sy;

    while (x != gx || y != gy)
    {
        cout << "(" << x << "," << y << ") -> ";

        if (x < gx && grid[x + 1][y] != 1)
            x++;
        else if (x > gx && grid[x - 1][y] != 1)
            x--;
        else if (y < gy && grid[x][y + 1] != 1)
            y++;
        else if (y > gy && grid[x][y - 1] != 1)
            y--;
        else
        {
            cout << "Blocked! No path found." << endl;
            return;
        }
    }

    cout << "(" << gx << "," << gy << ")" << endl;
}

int main()
{
    int grid[3][3] = {
        {0, 0, 0},
        {0, 1, 0}, // 1 is obstacle
        {0, 0, 0}};

    aStar(grid, 0, 0, 2, 2);

    return 0;
}
